From 81b6016557bbf2d508be73ac48684e0a5354ec9b Mon Sep 17 00:00:00 2001
From: Keyon Jie <yang.jie@linux.intel.com>
Date: Wed, 10 Jan 2018 16:59:15 +0800
Subject: [PATCH 1363/2367] ASoC: tdf8532: several fixes for SOF

1. work around the stop_play() failed issue to let next
playback works.
2. initialize the format setting with 2 channel, 16bit.

Signed-off-by: Keyon Jie <yang.jie@linux.intel.com>
---
 sound/soc/codecs/tdf8532.c | 52 ++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 52 insertions(+)

diff --git a/sound/soc/codecs/tdf8532.c b/sound/soc/codecs/tdf8532.c
index 1860e82..ccad958 100644
--- a/sound/soc/codecs/tdf8532.c
+++ b/sound/soc/codecs/tdf8532.c
@@ -228,6 +228,7 @@ static int tdf8532_stop_play(struct tdf8532_priv *tdf8532)
 		goto out;
 
 	ret = tdf8532_wait_state(tdf8532, STATE_STBY, ACK_TIMEOUT);
+
 	if (ret < 0)
 		goto out;
 
@@ -238,6 +239,13 @@ static int tdf8532_stop_play(struct tdf8532_priv *tdf8532)
 	ret = tdf8532_wait_state(tdf8532, STATE_IDLE, ACK_TIMEOUT);
 
 out:
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL)
+	/*
+	 * FIXME: this is only work around to fix the stop failed issue on SOF,
+	 * return success always to make the next start successfully.
+	 */
+	ret = 0;
+#endif
 	return ret;
 }
 
@@ -309,6 +317,8 @@ static int tdf8532_i2c_probe(struct i2c_client *i2c,
 	int ret;
 	struct tdf8532_priv *dev_data;
 	struct device *dev = &(i2c->dev);
+	u8 payload[16];
+	int ret1;
 
 	dev_dbg(&i2c->dev, "%s\n", __func__);
 
@@ -332,6 +342,48 @@ static int tdf8532_i2c_probe(struct i2c_client *i2c,
 		goto out;
 	}
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_INTEL)
+	/* We support 2 channels, 16bit 48KHz for SOF only at the moment */
+	dev_data->channels = 2;
+	/* set format */
+	payload[0] = 0x02;
+	payload[1] = 0x00;
+	payload[2] = 0x08;
+	payload[3] = 0x80;/* ModuleId */
+	payload[4] = 0x1c;/* CmdId, SetAudioConfig_cmd */
+	payload[5] = 0;/* rfa */
+	payload[6] = 0x01;/* SamplingFrequency, 48KHz */
+	payload[7] = 0x00;/* FrameSize, 2 slots per I2S frame */
+	payload[8] = 0x00;/* SlotSize, 16bit */
+	/* TDMFrameSync, TDM frame starts at rising edge frame sync */
+	payload[9] = 0x1;
+	payload[10] = 0x1;/* BitClockDelay, no delay */
+	ret1 = i2c_master_send(dev_data->i2c, payload, 11);
+
+	/* set routing */
+	payload[0] = 0x02;
+	payload[1] = 0x00;
+	payload[2] = 0x06;
+	payload[3] = 0x80;/* ModuleId */
+	payload[4] = 0x28;/* CmdId, SetAudioRoutingConfig_cmd */
+	/* channel 0 */
+	payload[5] = 0;/* AudioSource, SDI1 */
+	payload[6] = 0x00;/* SlotNr, Slot 0 */
+	/* channel 1 */
+	payload[7] = 0;/* AudioSource, SDI1 */
+	payload[8] = 0x01;/* SlotNr, Slot 1 */
+	ret1 = i2c_master_send(dev_data->i2c, payload, 9);
+
+	/* initial to disable all channels */
+	ret = tdf8532_amp_write(dev_data, SET_CHNL_DISABLE,
+			CHNL_MASK(8));
+
+#endif
+
+	if (ret < 0)
+		dev_err(&i2c->dev,
+			"i2c send set format packet returned:%d\n", ret);
+
 out:
 	return ret;
 }
-- 
2.7.4

